src/staking/block_reward_validator.cpp

```diff
@@ -15,27 +15,56 @@ namespace staking {
 class BlockRewardValidatorImpl : public BlockRewardValidator {
  private:
   Dependency<blockchain::Behavior> m_behavior;
+  Dependency<proposer::FinalizationRewardLogic> m_finalization_reward_logic;
 
  public:
   BlockRewardValidatorImpl(
-      Dependency<blockchain::Behavior> behavior)
-      : m_behavior(behavior) {}
+      Dependency<blockchain::Behavior> behavior,
+      Dependency<proposer::FinalizationRewardLogic> finalization_reward_logic)
+      : m_behavior(behavior),
+        m_finalization_reward_logic(finalization_reward_logic) {}
 
   bool CheckBlockRewards(const CTransaction &coinbase_tx, CValidationState &state, const CBlockIndex &index,
                          CAmount input_amount, CAmount fees) const override {
     assert(MoneyRange(input_amount));
     assert(MoneyRange(fees));
 
+    const CBlockIndex &prev_block = *index.pprev;
     CAmount total_reward = fees + m_behavior->CalculateBlockReward(index.nHeight);
 
-    std::size_t num_reward_outputs = 1;
+    std::size_t num_reward_outputs = m_finalization_reward_logic->GetNumberOfRewardOutputs(index.nHeight) + 1;
     if (coinbase_tx.vout.size() < num_reward_outputs) {
       return state.DoS(100,
                        error("%s: too few coinbase outputs expected at least %d actual %d", __func__,
                              num_reward_outputs, coinbase_tx.vout.size()),
                        REJECT_INVALID, "bad-cb-too-few-outputs");
     }
 
+    if (num_reward_outputs > 1 && !(prev_block.pprev->nStatus & BLOCK_HAVE_DATA)) {
+      // prev_block is a parent block of the snapshot which was used for ISD.
+      // We do not have data for the ancestor blocks of prev_block.
+      // TODO UNIT-E: implement proper validation of finalization rewards after ISD
```

Comment:
I'd like to have an issue for this. It seems that we cannot check if the rewards are being payed to the correct scripts here, correct?

Reply:
Did we create an issue for this?